### ISSUE TITLE
Using volumes via network (SMB,iSCSI) makes it slow - way slower than regular SMB,iSCSI access

### ISSUE THREAD

**Issue by kstepyra** [veracrypt/VeraCrypt#1277]
I am connected to 8 drive NAS via 10gbe, when i copy to it i can top out >1GB/s of transfer. Whenever i try to copy anything to/from VC container it tops out at about 100-150MB/s. Changing NTFS block size doesn't change anything.
Tried changing everything - network stack on TrueNAS, all filesystem types/block sizes on VC container, system on main machine - nothing helps. Last thing i didn't check is newer VC version - but i don't see anything related in changelog. 

# Expected behavior
Veracrypt shouldn't slow down drastically when accessing VC container via network (10% of possible speed here)

# Observed behavior
Copying files from/to mounted VC over SMB/iSCSI is very slow

# Steps to reproduce
1. Write/read from NAS (SMB/iSCSI) to benchmark its speed
2. Create VC container (any filesystem type) on NAS via SMB/iSCSI, with AES for best enc/dec speed on modern CPU
3. Write/Read files from it on Windows 10 or Ubuntu 22 server machine
4. Don't benefit :)

# Your Environment
NAS server is on TrueNAS ZFS array (raidz1) with 128KB block size, connected using 10gbit nic
Main machine is either ubuntu or windows 10, both with 5950x, 128GB of RAM and samsung 970 evo plus 2TB drive.

**VeraCrypt version:** 1.25.9 x64

**Operating system and version:** Windows 10 version 22H2 build 19045.3693

**System type:**  64-bit

**Comment by nerai** [comment]
If you have an SSD in the machine, can you test the R/W speed with a VC container or partition there? This should help to determine if the medium being a RAID is a factor.

**Comment by kstepyra** [comment]
Directly on TrueNAS server, without client machine and SMB?

**Comment by thany** [comment]
Same problem here. It oftenly grinds to below 1MB/s even though the target server has modern disks that are otherwise idle and capable of pullong 200MB/s. It's also on 2.5Gb ethernet that is also otherwise idle. Safe to say, in my case, the server or my connection to it, is absolutely not the problem.

I did notice that VC makes the server work REALLY HARD, by somehow writing super slowly but maybe in random places or something. Not a problem for SSDs, but a huge pita for HDDs. Don't know if this is the case, but again, dragging a huge regular file to the same SMB share writes it at full blast without problems.

Moreover, while this grinding to a halt is happening, the local pc running VC also grinds. Many I/O based operations are delayed, leading to huge lags all over the system. Even something as simple as opening task manager took almost a minute. This is absolutely unacceptable on the kind of system I have, and it's completely VC's fault, because the pc is otherwise as fast as a mortal person can have in this day and age.

And the fact that VC happily "bursts" or "buffers" writes *well* in excess of the target volume's maximum speed, makes the problem even worse. This means it will keep grinding and grinding and grinding well after having sucessfully cancelled a copy operation to the VC volume.

**Comment by thany** [comment]
> It would perhaps also be interesting to see which type of RAID (hardware, mdadm, zfs, ...), which level (0/1/5/6/...), which block size etc. I've tried settings that should be advantageous for VC but the improvements were insufficient.

For me it's [Unraid](https://unraid.net/). My VC containers are on the main array, consisting of 4 disks, one of which is parity. This is not exactly RAID (that's the whole point of Unraid), but it does keep parity. So perhaps it's something to do with the block sizes that VC writes, or block alignment, or something along those lines. Probably the reason why reading performs normal, afaict, but writing speed degrades pretty quickly.

But this still doesn't address VC over-willingness to burst writes, to sort "take" a large chunk of write operations up to a point, at incredible speed, well in excess of whatever the physical volume file is stored on. This bursting behaviour certainly doesn't help in this scenario.

**Comment by amadboss** [comment]
> The relevant points, as far as I can tell:
> 
> * VC on a local machine is fast, for both SSD and HDD (100 MB/s or more). This rules out the CPU as a culprit.
> * VC on a remote drive that is NOT in a RAID, accessed via SMB, is fast (?) or at least fast-ish (50+ MB/s for large file transfers). This rules out the RAID as a definite/unambiguous bottleneck.
> * VC on a remote drive that IS in a RAID of some kind, accessed via SMB,  can be slow (3 MB/s or so).
> * The same remote drive in a RAID, accessed via SMB, WITHOUT VC, is fast (100+ MB/s). This rules out the connection and the top speed of the remote drive as culprits.
> 
> The important part seems to be the combination of VC, SMB and RAID. If any of these factors is omitted, it works fine. I've confirmed that TrueCrypt behaves identically in these cases.
> 
> It would be great if others could confirm if this is the case for them, too, or if their problem requires other/fewer factors.
> 
> It would perhaps also be interesting to see which type of RAID (hardware, mdadm, zfs, ...), which level (0/1/5/6/...), which block size etc. I've tried settings that should be advantageous for VC but the improvements were insufficient.
I personaly use OMV with 4 disk in raid 5 (they use mdadm for RAID) and i have exactely the same issue as mentioned. 
With Raid + smb + vc my netword speed drop drop ~1Gbps (with raid + smb) to 40Mbps (raid+smb+vc). 
I test pretty mush all settings on vc container like buffer size, file system or encryption algorithm but nothing change.

Does anyone have some news about this issue on how to solve and why is this happening ?

**Comment by thany** [comment]
> Does anyone have some news about this issue on how to solve and why is this happening ?

Yeah I was hoping the veracrypt folks would have replied by now. At least something like "we're on it" would be more affirming than nothing at all.

**Comment by nerai** [comment]
> > Does anyone have some news about this issue on how to solve and why is this happening ?
> 
> Yeah I was hoping the veracrypt folks would have replied by now. At least something like "we're on it" would be more affirming than nothing at all.

This issue is quite hard to replicate because it requires specific hardware and, apart from that, I'm sure they have a life of their own, so this may take some time. We should wait patiently :)

**Comment by thany** [comment]
@nerai We should, but it can't hurt if the developer could post at least a quick status update, rather than leaving us "in the dark", so to speak. That way, at the very least, it's an indication that it's on their todo list.

**Comment by thany** [comment]
It's 3 months later, and an update would be both helpful and greatly appreciated.

I'm seriously not sure what the solution is. I don't even know how to diagnose any of this, as it's quite "low level" and therefor hard to "get to" from userland.

I'm happy to try out some stuff, but I don't know what 😞

**Comment by vedimh** [comment]
Hi, I can confirm this issue and add some additional details.
I have four computers in local network (1Gbps) - two Windows 10 and two Debian 12 (dual boot Windows 10).

When I mount Veracrypt file from NTFS shared drive on Debian 12 and try to copy some larger file to it (upload from Windows 10 to Debian), transfer speed never exceeds 60-70mbps. But, if copy something in opposite direction (download from Debian to Windows 10), I get around 700Mbps. This happens with both Debian PCs. 

Also, if I reboot those two back to Windows 10 and then mount Veracrypt over shares again, I get over 600Mbps in both directions.

I should also mention that copying files between Windows and Debian (no Veracrypt mounts) works with full 1Gbps in both directions.

Any help greatly appreciated :)

**Comment by IntCleastwood** [comment]
Same here in combination with MacOS Sonoma + Unraid NAS  connected via 2.5 Gbit. I am connected via WiFi6 and sitting near to the router, where iperf3 shows me approx 2 - 2,1 GBit for general network speed between NAS + Macbook.

Following different threads, especially samba + MacOS seems to be an unlucky combination, but i can gain up to 180MBit for a general file copy from my Macbook to the NAS share (no encryption involved).

Creating a container directly on the samba share drops to 100MB/s which is okayish/expected, since its only a one-timer.

But the mounted container via SMB share is painfully slow, about 20-30MB/s which makes it unusable for example Photo editing.

**Comment by thany** [comment]
I've moved a container on Unraid from the main array (which is spinning rust) to a RAID-5 ZFS array (which is SSD all the way). Mounting that on Windows via an SMB share yields in the same performance, which doesn't make any sense. Surely, 4 SSD's raided together should *easily* be able to saturate the 2.5Gb wired networking between both.

And also, its bursting behaviour makes it *even* slower, by faking high speed first and then having to write back all that data at reduced "real" speed.

**Comment by vedimh** [comment]
@nerai No, at least not for me. For me, the problem appears only when mounting Veracrypt volume hosted on Linux to Windows PC. In this case transfer from Windows to Linux (copying file to Veracrypt volume) is extremely slow, but transfer in opposite direction is just fine. Also, all other mounting combinations work without issues - local mounts, Windows to Windows, Linux to Linux and Linux to Windows. It is really,really weird.

**Comment by thany** [comment]
I wonder if we can disable write bursting, or whatever it's called, so that copy operations can be tested for speed more easily and genuinely.

With VC bursting writes, it's really difficult to see if a copy to a volume is genuinely fast, or if it'll grind down after "taking" a few hundred MB.

**Comment by nerai** [comment]
A few 100MB? For me the burst buffer seems to be about 10GB. I have 128GB RAM, maybe that's influencing it.

**Comment by thany** [comment]
Veracrypt devs: Any progress?

Would be nice to get an update on what you've been doing so far about this problem.

**Comment by glaserf** [comment]
Can confirm, performance is still a disaster.

**Comment by kstepyra** [comment]
> **Optimization Advice:** Implement local caching layers (like Windows BranchCache or application-level caching) and optimize network parameters (MTU size, TCP window scaling, SMB3 multichannel) to reduce latency and increase throughput for remote volume access. Additionally, consider tiering frequently accessed data to local storage or using dedicated high-bandwidth, low-latency network infrastructure (10GbE+, dedicated VLAN) to minimize the performance gap between local and network-attached volumes.
> 
> **Metrics Question:** Are you measuring latency (ms), throughput (MB/s), and IOPS separately, and have you compared these metrics between direct SMB/iSCSI access versus volume-mounted access to isolate whether the bottleneck is network, protocol overhead, or the volume abstraction layer itself?

Every other app i used could saturate that 10gbit, network is totally fine. yes, i changed every option in kernel, as well in windows what could be changed. I tested linux<->linux and every other combination. SMB3 multichannel was also tested. " frequently accessed data" - i do bulk transfers of different data, thats why i need bandwidth. It's still VC problem, not any other - just do a tests yourself and you will see. Show me a transfer of 1GB/s on LAN to VC volume if it's possible.

**Comment by nerai** [comment]
@Solix45456767 Based on the discussion, I'd recommend to ban bot accounts from suggesting crap ideas that waste the time of human contributors who are inundated in doing free labor.

**Comment by kstepyra** [comment]
> [@Solix45456767](https://github.com/Solix45456767) Based on the discussion, I'd recommend to ban bot accounts from suggesting crap ideas that waste the time of human contributors who are inundated in doing free labor.

So you say instead of fixing a bug that makes it almost unusable in a lot of use cases (NAS is nothing new for VC containers), just run away with resolution that devs don't have time to contribute to open source project they already started to work on? Then why even start such a project? Just build commercial one, if money is only thing that matters. You probably type that on linux machine that is built by someone else, also for free, why not build kernel software all day long?

**Comment by nerai** [comment]
> > [@Solix45456767](https://github.com/Solix45456767) Based on the discussion, I'd recommend to ban bot accounts from suggesting crap ideas that waste the time of human contributors who are inundated in doing free labor.
> 
> So you say instead of fixing a bug that makes it almost unusable in a lot of use cases (NAS is nothing new for VC containers), just run away with resolution that devs don't have time to contribute to open source project they already started to work on? Then why even start such a project? Just build commercial one, if money is only thing that matters. You probably type that on linux machine that is built by someone else, also for free, why not build kernel software all day long?

Misunderstanding. My comment was about @Solix45456767, who is a bot that posts bullshit. This wastes our (human's) time. I'm a maintainer of open source software myself, so I know how infuriating that is. I still hope that the devs here can find the time and resources to fix the original issue, and wasting their time will certainly not help.

I've reported the bot and it has been deleted.

**Comment by thany** [comment]
In the mean time, I've migrated my volumes to BitLocker volumes. This is unlikely a solution for those working with Linux (as the client) but if you're on Windows, it's worth considering. BL volumes can saturate the network if mounted via SMB, just like any other network traffic. And by securing them with just a password, they remain portable as well (even between two different Windows versions). Securing them with a Windows account probably breaks their portability. I haven't yet had the courage to secure them with biometrics.

I've also written a powershell script to mount and unlock all of them at once, having to supply the password only once. Just like how VC works when caching passwords is enabled.

It just sucks that VC can't achieve any respectable performance over the network. It's a really rather weird problem. It would make my volumes more portable (e.g. to Linux) and give more control (algorithms and such). But if it doesn't work, it shouldn't be a surprise that users start looking for alternate solutions. I hope the VC devs realise that.

**Comment by adelin-mcbsoft** [comment]
@thany , just one mention for BitLocker volumes:

While they do the job indeed (and I use them daily too), there's an ugly thing I dislike about them -- and that's the fact that you can't disable write caching (to make it `direct-sync` like).

i.e. You start copying some files to it, Windows says it finished copying, but in the background it still copies data over the network.
If for some reason your connection breaks or you eject the volume / shutdown / whatever, you get corrupted (or loss of) data.

That's why I wanted to switch to VeraCrypt, but indeed, performance is drastically reduced over SMB.

**Comment by thany** [comment]
@adelin-mcbsoft You're right. Well, probably. I can't check right now, but I think there's a setting at some point for it. But unmounting/remounting a volume likely resets that setting. It would be great if there was a flag for write caching in the powershell command to mount one, but I don't think there is. So I think you're right.

Bummer.

But between these two options, VC's ridiculous performance is far more annoying, since my network is stable. I wouldn't do it over wifi, safe to say 🙂

**Comment by TETYYS** [comment]
I've managed to get it working significantly faster by adding `strict sync = no` to smb.conf

**Comment by thany** [comment]
> I've managed to get it working significantly faster by adding `strict sync = no` to smb.conf

It shouldn't require server config. BitLocker is working at full speed, so VeraCrypt should be able, too.

**Comment by TETYYS** [comment]
yeah, but as proper solution doesn't exist, it might be next best thing

**Comment by nerai** [comment]
> I've managed to get it working significantly faster by adding `strict sync = no` to smb.conf

I can confirm that this has improved the situation for me greatly. Went from 3MB/s and freezing the client machine to 50MB/s with no freezing. The network/disk themselves can sustain 250MB/s, but I'll be satisfied.

Perhaps VC flushes very often, so this could be an issue. It would be nice to have more information about the typical access pattern that is expected of the program, so we can try to mitigate it further.
